src/useBlocklyWorkspace.js

```diff
@@ -0,0 +1,130 @@
+import React from "react";
+import Blockly from "blockly";
+
+import debounce from "./debounce";
+
+function importFromXml(xml, workspace, onImportXmlError) {
+  try {
+    Blockly.Xml.domToWorkspace(Blockly.Xml.textToDom(xml), workspace);
+    return true;
+  } catch (e) {
+    if (onImportXmlError) {
+      onImportXmlError(e);
+    }
+    return false;
+  }
+}
+
+const useBlocklyWorkspace = ({
+  ref,
+  initialXml,
+  toolboxConfiguration,
+  workspaceConfiguration,
+  onWorkspaceChange,
+  onImportXmlError,
+}) => {
+  const [workspace, setWorkspace] = React.useState(null);
+  const [xml, setXml] = React.useState(initialXml);
+  const [didInitialImport, setDidInitialImport] = React.useState(false);
+  const [didHandleNewWorkspace, setDidHandleNewWorkspace] = React.useState(false);
+
+  // we explicitly don't want to recreate the workspace when the configuration changes
+  // so, we'll keep it in a ref and update as necessary in an effect hook
+  const workspaceConfigurationRef = React.useRef(workspaceConfiguration);
+  React.useEffect(() => {
+    workspaceConfigurationRef.current = workspaceConfiguration;
+  }, [workspaceConfiguration]);
+
+  const toolboxConfigurationRef = React.useRef(toolboxConfiguration);
+  React.useEffect(() => {
+    toolboxConfigurationRef.current = toolboxConfiguration;
+    if (toolboxConfiguration && workspace) {
+      workspace.updateToolbox(toolboxConfiguration);
+    }
+  }, [toolboxConfiguration, workspace]);
+
+  const handleWorkspaceChanged = React.useCallback(
+    (newWorkspace) => {
+      if (onWorkspaceChange) {
+        onWorkspaceChange(newWorkspace);
+      }
+    },
+    [onWorkspaceChange]
+  );
+
+  // Workspace creation
+  React.useEffect(() => {
+    const newWorkspace = Blockly.inject(ref.current, {
+      ...workspaceConfigurationRef.current,
+      toolbox: toolboxConfigurationRef.current,
+    });
+    setWorkspace(newWorkspace);
+    setDidInitialImport(false); // force a re-import if we recreate the workspace
+    setDidHandleNewWorkspace(false); // Singal that a workspace change event needs to be sent.
+
+    // Dispose of the workspace when our div ref goes away (Equivalent to didComponentUnmount)
+    return () => {
+      newWorkspace.dispose();
```

Comment:
some plugin also may register cleanup like keyboard navigation

Reply:
Good point; I'll add an onDispose while I'm in here.